### ISSUE TITLE
homeassistant and openhab can only read but not write

### ISSUE THREAD

**Issue by p4jo** [knxd/knxd#539]
Using the integrations for `knxd` in the `homeassistant` and `openhab` Smart Home systems, I cannot get it to work that changing the state of an item (e.g. turning a switch on or off) actually does something in `knxd`.
However, they recieve and show changes that happen on the KNX bus correctly (through the multicast address). So, if I click a switch that's in the bus, they immediately reflect the change of state.
They also do send signals to the multicast address 224.0.23.12 that look alsmost identical (using `tcpdump`) to the signals generated by `knxtool`. The `knxtool on ip:localhost 4/2/0` calls work flawlessly.

## Details and logs

* `openhab` config:
```
UID: knx:ip:087ebd7566
label: KNX/IP Gateway
thingTypeUID: knx:ip
configuration:
  useNAT: false
  readRetriesLimit: 3
  autoReconnectPeriod: 60
  localIp: 192.168.179.42
  localSourceAddr: 1.1.105
  readingPause: 50
  type: ROUTER
  responseTimeout: 10
  portNumber: 3671
```
* `homeassistant` config: 
![grafik](https://user-images.githubusercontent.com/44933755/187026096-a2e57616-ca8c-47de-92f2-6e8829bb78c4.png)

* `knxd` config:
```
[main]
addr = 1.1.104
client-addrs=1.1.105:5
connections = server,A.tpuarts
# ,A.tcp,ip
# cache = A.cache
systemd = systemd
[server]
debug = debug-server
discover = true
router = router
server = ets_router
tunnel = tunnel
# multicast-address=224.0.23.12
# port=3671
[debug-server]
name = mcast:knxd
[A.tpuarts]
device = /dev/knxd1
# baudrate = 9600
driver = tpuart
ack-group = true
ack-individual = true
# [A.tpuarttcp]
# driver=tpuarttcp
# ip-address=192.168.179.42
# dest-port=55332
# [A.tcp]
# server=knxd_tcp
# systemd-ignore=true
# port=6722
# [ip]
# driver=ip
# multicast-address=224.0.23.13
[systemd]
debug = debug-systemd
filters = log
[debug-systemd]
error-level=0x9
trace-mask=0xfff
```
### Using `sudo tcpdump -A port 3671`.
* Signal that `knxd` generates when changing the light at 4/2/0 with the bus switch: 
```
13:02:14.465600 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-..@...jQ...*.....W.W..k
...0..)....."....
```
Log in `journalctl -u knxd -f`:
```
Aug 27 13:02:14 openhabian knxd[546828]: Layer 6 [ 9:systemd_/systemd      1938.767] is OK
Aug 27 13:02:14 openhabian knxd[546828]: Layer 6 [ 4:systemd_/systemd      1938.767] is OK
```

* Signal that `knxtool on/off ip:localhost 4/2/0` generates:
```
12:38:55.253191 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-..@........*.....W.W..k
...0..)....k"....
12:39:13.298442 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-.F@...
....*.....W.W..k
...0..)....l"....+

[...]

12:59:06.225118 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-].@........*.....W.W..k
...0..)....j"....
```
Log of it (the last one):
```Aug 27 12:59:06 openhabian knxd[546828]: Layer 8 [ 9:systemd_/systemd      1750.505] New Connection
Aug 27 12:59:06 openhabian knxd[546828]: Layer 8 [59:systemd_/CConn        1750.505] ClientConnection Init
Aug 27 12:59:06 openhabian knxd[546828]: Layer 3 [59:systemd_/CConn        1750.505] Allocate 1.1.106
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [59:systemd_/CConn        1750.505] ReadMessage(005): 00 22 22 00 FF
Aug 27 12:59:06 openhabian knxd[546828]: Layer 7 [60:systemd_@1.1.106/Group 1750.505] OpenGroup
Aug 27 12:59:06 openhabian knxd[546828]: Layer 4 [62:systemd_/TGr          1750.505] OpenGroup 4/2/0 WO
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [63:log/systemd_          1750.505] State setup
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [59:systemd_/CConn        1750.506] Send(002): 00 22
Aug 27 12:59:06 openhabian knxd[546828]: Layer 7 [60:systemd_@1.1.106/Group 1750.506] OpenGroup complete
Aug 27 12:59:06 openhabian knxd[546828]: Layer 3 [61:systemd_/ConnS        1750.506] registerLink: 61:systemd__61
Aug 27 12:59:06 openhabian knxd[546828]: Layer 3 [61:systemd_/ConnS        1750.506] Start: cfg:systemd_
Aug 27 12:59:06 openhabian knxd[546828]: Layer 5 [61:systemd_/ConnS        1750.506] down => >up
Aug 27 12:59:06 openhabian knxd[546828]: Layer 5 [61:systemd_/ConnS        1750.506] Starting
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [63:log/systemd_          1750.506] State start
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [63:log/systemd_          1750.506] State started
Aug 27 12:59:06 openhabian knxd[546828]: Layer 5 [61:systemd_/ConnS        1750.506] >up => up
Aug 27 12:59:06 openhabian knxd[546828]: Layer 4 [61:systemd_/ConnS        1750.506] link state changed: up
Aug 27 12:59:06 openhabian knxd[546828]: Layer 5 [61:systemd_/ConnS        1750.506] Started
Aug 27 12:59:06 openhabian knxd[546828]: Layer 4 [61:systemd_/ConnS        1750.506] link state changed: up
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [59:systemd_/CConn        1750.506] ReadMessage(004): 00 25 00 80
Aug 27 12:59:06 openhabian knxd[546828]: Layer 7 [59:systemd_/CConn        1750.507] recv Group(002): 00 80
Aug 27 12:59:06 openhabian knxd[546828]: Layer 4 [62:systemd_/TGr          1750.507] Recv Group T_Data_Group A_GroupValue_Write (small) 00
Aug 27 12:59:06 openhabian knxd[546828]: Layer 0 [63:log/systemd_          1750.507] Recv L_Data low from 0.0.0 to 4/2/0 hops: 06 T_Data_Group A_GroupValue_Write (small) 00
Aug 27 12:59:06 openhabian knxd[546828]: Layer 8 [61:systemd_/ConnS        1750.507] found addr 1.1.106
Aug 27 12:59:06 openhabian knxd[546828]: Layer 8 [59:systemd_/CConn        1750.507] ClientConnection 1.1.106 closing
Aug 27 12:59:06 openhabian knxd[546828]: Layer 3 [61:systemd_/ConnS        1750.507] unregisterLink: systemd__61
Aug 27 12:59:06 openhabian knxd[546828]: Layer 8 [59:systemd_/CConn        1750.507] Exiting
Aug 27 12:59:06 openhabian knxd[546828]: Layer 7 [59:systemd_/CConn        1750.507] CloseGroup
Aug 27 12:59:06 openhabian knxd[546828]: Layer 4 [62:systemd_/TGr          1750.507] CloseGroup
Aug 27 12:59:06 openhabian knxd[546828]: Layer 6 [ 9:systemd_/systemd      1750.534] is OK
Aug 27 12:59:06 openhabian knxd[546828]: Layer 6 [ 4:systemd_/systemd      1750.534] is OK
```

* Signal that flipping the switch in `homeassistant` generates:
```
12:44:51.658903 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-.n@...Sr...*.....W.W..k
...0..)....i"....
12:44:55.498529 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-.>@...R....*.....W.W..k
...0..)....i"....
```
Log in `journalctl -u knxd -f`:
``` ```
* Signal that flipping the switch in `openhabian` generates
```
12:46:30.931684 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-..@.@......*.....W.W..k
...0..)....i"....
12:46:32.710659 IP 192-168-179-42.fritz.box.3671 > 224.0.23.12.3671: UDP, length 17
E..-..@.@..Z...*.....W.W..k
...0..)....i"....
```
Log in `journalctl -u knxd -f`:
``` ```

## Other ideas
I also tried the method described in the updated `inifile.rst`, using `socat`: installing `knxd` on another Raspberry pi in my local network to avoid the packets being filtered out because they come from the same address that recieves it in multicast. It didn't change anything. 
When a signal from `knxtool` came, a signal was sent through the `socat` link to my bus (`/dev/knxd1` on my original Pi 4), but not when I clicked in one of the smart home systems. (as seen by `sudo tcpdump -A port 55332`), otherwise it's identical to what I described above.

`socat TCP-LISTEN:55332,reuseaddr /dev/knxd1,b19200,parenb,raw`
* `knxd` config on my other pi (at `Pi-Keller.fritz.box`) -- the original one is at `192.168.179.42`/`openhabian`/`pi-openhab`)
```
[main]
addr = 1.1.104
client-addrs=1.1.105:5
connections = server,A.tpuarts
systemd = systemd
[server]
debug = debug-server
discover = true
router = router
server = ets_router
tunnel = tunnel
# multicast-address=224.0.23.12
# port=3671
[debug-server]
name = mcast:knxd
[A.tpuarts]
#baudrate=9600
driver=tpuart
ip-address=192.168.179.42
dest-port=55332
[systemd]
debug = debug-systemd
filters = log
[debug-systemd]
error-level=0x9
trace-mask=0xfff
```
### other configs
I have tried almost everything in terms of changing the config (according to the `inifile.rst`) as you can maybe tell by the commented lines above. I also exhausted all possibilities of configs in the smart home systems.

I am very frustrated by this because it has been like that for a long time and the alternative would be 
 * to essentially write a program that recieves the multicast calls and makes `knxtool` calls out of them. This is terrible because I would essentially reprogram parts of `knxd`; 
 * or to do some crazy setup in the smart home systems, somehow calling `knxtool`. However the automation capabilities aren't so great and calling the system console is hard as well.

## TL;DR
`knxd` filters out the write requests that smart home integration systems send to the multicast address. It doesn't even show a log like "dropped".
However it flawlessly works in the other direction, sending changes on the bus to the multicast address from where the smart home systems recieve them correctly.
Using `knxtool` works flawlessly, too. It turns the lights on/off and also the smart home systems track the change.

**Comment by ozett** [comment]
seems also my problem...
cannot send multicast to knxd on the same raspberry from node-red.
hope to get some hints here how to debug further
(knxd runs fine with sending to to bus if multcast comes from other maschines)

**Comment by smurfix** [comment]
If you want to send multicast packets to other programs on the same host you need to set a flag on the **sending** socket. There's nothing knxd can do to receive these packets if you don't do that. Please file an issue with the sender.

Alternately, use tunnel mode.

**Comment by smurfix** [comment]
cf. line 114 ff. in `src/libserver/eibnetip.cpp`:
```
    char loopch=1;
    setsockopt(fd, IPPROTO_IP, IP_MULTICAST_LOOP, &loopch, sizeof(loopch));

```

**Comment by smurfix** [comment]
Also if you use routing you **must not** use addresses from knxd's client space. Those are strictly for tunnels and `knxtool`!

**Comment by p4jo** [comment]
Thank you for your response!
I sadly can't look into it until next weekend, but I'll try to bring this forward to the respective authors if I can't get to fix it.

The addresses of the switches, lights, etc. in the home network are all different from the ones I specified for knxd in the ini file. 

I still wonder why the issue wasn't fixed by rerouting using socat.

I've sadly not gotten tunnel mode to work for me, but that may be because I don't really understand it.

**Comment by ozett** [comment]
> you need to set a flag on the **sending** socket.

hi, thanks for juming in. seems all also my problem.
can i solve this on the raspberry/linux level? "set a flag on the sending socket" ?
or setting on knx-level ? is it something for the linx-multicast-configuration, or must this be done in the node-red module? 

thanks for clarification, i am hoping to let node-red run on the same device togehter with knxd and multicast.
but maybe my understanding is not deep enough to get this running...

**Comment by smurfix** [comment]
> can i solve this on the raspberry/linux level? "set a flag on the sending socket" ?

That's the job of the **program** (node-red / openhab / …) that sends the packet. There's nothing you can do in configuration or whatever, this has to be solved in the code.

**Comment by p4jo** [comment]
> Also if you use routing you **must not** use addresses from knxd's client space. Those are strictly for tunnels and `knxtool`!

This has solved the issue! I wouldn; ever have thought of explicitly specifying a different address in my client applications that what I used in knxd's settings.

It still only works when running on a different machine. However I decided to actually connect the other pi to the KNX bus rather than relaying everything via network there (and back via multicast).

Thanks for saving our smart home installation!
